### PR TITLE
Replace cast+multiply masking with where in forward-pass layers

### PR DESCRIPTION
**Pull request by Saumay**
## Description

Fixes part of #22386.

`tensor * ops.cast(bool_mask, tensor.dtype)` materializes a full float copy of the mask just to multiply by it. JAX and TF fuse this away during compilation, but torch runs it eagerly, so the intermediate sticks around and can push big layers into OOM.

I switched the forward-pass sites listed in the issue to use `where(mask, tensor, 0)` instead, which keeps the mask as a bool:

- `layers/attention/attention.py`: padding-mask bias and the q_mask zero-out after attention.
- `layers/normalization/batch_normalization.py`: reworked masked `_moments` so the weighted sums consume short-lived `where` results instead of holding a broadcasted mask, weighted inputs, difference, and squared difference all at once.
- `layers/core/masking.py`: post-mask zero step.
- `layers/pooling/global_average_pooling1d.py`: masked numerator. While I was there, this also stops `inputs *= mask` from mutating the caller's input tensor under the torch backend.
- `activations/activations.py`: ReLU with a custom threshold.
- `constraints/constraints.py`: `NonNeg`.

In the BN, GAP1D, and attention-q_mask sites I also added a defensive `ops.cast(mask, "bool")` on the user-supplied mask before it reaches `where`, matching the pattern already used in `MultiHeadAttention` and `GroupedQueryAttention`. Without it the bool `where` and the float `sum_of_weights` cast would disagree on non-binary masks.

Losses/metrics sites in #22386 are smaller. I'll send a follow-up for those.

### Tests

pytest passes on tensorflow, jax and torch (CPU) for each touched module:

- `keras/src/layers/attention/attention_test.py`
- `keras/src/layers/normalization/batch_normalization_test.py`
- `keras/src/layers/core/masking_test.py`
- `keras/src/layers/pooling/global_average_pooling_test.py`
- `keras/src/activations/activations_test.py`
- `keras/src/constraints/constraints_test.py`

### AI-assisted disclosure

Per the AI-Assisted Contribution Policy, disclosing that I used Claude Code to help draft the rewrites and tests. I reviewed every change, ran the full test suite per backend locally, and take responsibility for the code.

## Contributor Agreement
**Please review our
[AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy)
and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/layers/normalization/batch_normalization.py [hunk 1]
```diff
@@ -439,29 +439,28 @@ def _moments(self, inputs, mask):
...
-
-        weighted_input_sum = ops.sum(
-            weighted_inputs,
+        # Compared to the prior cast+broadcast+multiply pattern, this keeps
+        # only short-lived inputs-sized intermediates (the `where`/broadcast
+        # results are released as soon as each `ops.sum` returns) instead of
+        # holding a full broadcasted mask + weighted inputs + difference +
+        # squared difference simultaneously. Reduces peak memory under the
+        # torch backend, which doesn't fuse these eagerly (see #22386).
+        mask_b = ops.expand_dims(mask, axis=-1)
```
[the remark this answers is not in the document]
Comment (Saumay) [reply keras/src/layers/normalization/batch_normalization.py]: This `axis=-1` was already on master before this PR; I just preserved it. Generalizing to non-last `self.axis` feels like a separate change to me.

File: keras/src/activations/activations.py [hunk 2]
```diff
@@ -90,9 +90,7 @@ def static_call(x, negative_slope=0.0, max_value=None, threshold=0.0):
         if threshold != 0:
             # computes x for x > threshold else 0
             threshold = ops.cast(threshold, dtype=x.dtype)
-            x = x * backend.cast(
-                backend.numpy.greater(x, threshold), dtype=x.dtype
-            )
+            x = backend.numpy.where(backend.numpy.greater(x, threshold), x, 0.0)
```
[the remark this answers is not in the document]
Comment (Saumay) [reply keras/src/activations/activations.py]: Kept `backend.numpy.where` so it stays consistent with the surrounding `backend.numpy.greater`, `backend.cast`, and `backend.numpy.clip` calls in this `static_call`. Took the `0.0` to `0` change though, thanks.

File: keras/src/layers/normalization/batch_normalization.py [hunk 3]
```diff
@@ -439,29 +439,28 @@ def _moments(self, inputs, mask):
...
+        mask_float = ops.cast(mask_b, inputs.dtype)
+        sum_of_weights = ops.sum(
+            ops.broadcast_to(mask_float, ops.shape(inputs)),
             self._reduction_axes,
             keepdims=True,
         )
```
[the remark this answers is not in the document]
Comment (Saumay) [reply keras/src/layers/normalization/batch_normalization.py:450]: I tried this first and it actually breaks `test_masked_broadcast_normalization`. The mask only covers a subset of the reduction axes (for example mask `(1, 2, 1)` with input `(1, 2, 3, 4)` and `_reduction_axes=[0, 1, 2]`), so without the broadcast the sum loses the factor from the axes the mask does not span and `sum_of_weights` ends up too small. The broadcast result is consumed inside `ops.sum` and freed right after, so keeping it does not affect peak memory.

**Comment by hertschuh** [comment]
This was covered by https://github.com/keras-team/keras/pull/22392
